tracer/test/Datadog.Trace.TestHelpers/MockApi.cs

```diff
@@ -9,18 +9,36 @@
 using System.Threading;
 using System.Threading.Tasks;
 using Datadog.Trace.Agent;
+using FluentAssertions;
 
 namespace Datadog.Trace.TestHelpers
 {
     internal class MockApi : IApi
     {
-        private readonly ManualResetEventSlim _resetEvent = new();
-        private List<List<MockSpan>> _objects = null;
+        private readonly ManualResetEventSlim _resetEvent;
+        private readonly object _lock = new();
+        private List<List<MockSpan>> _objects = new();
 
-        public List<List<MockSpan>> Wait()
+        public MockApi(ManualResetEventSlim resetEvent = null)
         {
-            _resetEvent.Wait();
-            var objects = Interlocked.Exchange(ref _objects, null);
+            _resetEvent = resetEvent ?? new();
+        }
+
+        public List<List<MockSpan>> Traces
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    return _objects;
+                }
+            }
+        }
+
+        public List<List<MockSpan>> Wait(TimeSpan? timeout = null)
+        {
+            _resetEvent.Wait(timeout ?? TimeSpan.FromMinutes(1)).Should().BeTrue();
```

Comment:
I'm a little conflicted about adding an assert here because a failure here has the possibility of throwing an exception and not running any cleanup code. I'm not sure I have concrete suggestion, but it's a general thought I had

Reply:
An alternative could be to ignore this, return `objects`, and let the test assert the expected number of spans.

Reply:
> I'm a little conflicted about adding an assert here because a failure here has the possibility of throwing an exception and not running any cleanup code. 

Yeah, my initial thought was that cleanup code has to go in a `Dispose()` on the class for tests to be sure it runs anyway, so this wouldn't matter. And previously this would hang rather than throw which is definitely worse. That said, we take the approach Lucas suggests in `MockTracerAgent.WaitForSpans()`, so we should probably just do the same here.

(Though, ironically, I think _not_ throwing in `MockTracerAgent.WaitForSpans()` is the cause of some long running tests, where we have the wrong conditions, and so are exhausting the timeout every time 😉)